File: src/ts/maximumUniqueSubarray.ts
export function maximumUniqueSubarray (nums: number[]): number {
  let result = Number.MIN_SAFE_INTEGER
  let sum = 0
  const visits = new Map<number, number>()
  for (let left = 0, right = 0; right < nums.length; right += 1) {
    const rn = nums[right]
    if (visits.has(rn)) {
      const visited = visits.get(rn) as number
      while (left <= visited) {
        sum -= nums[left]
        left += 1
      }
    }

    sum += rn
    result = Math.max(result, sum)
    visits.set(rn, right)
  }

  return result
}

File: src/rs/maximum_unique_subarray.rs
pub struct MaximumUniqueSubarray {}
impl MaximumUniqueSubarray {
    pub fn maximum_unique_subarray(nums: Vec<i32>) -> i32 {
        use std::collections::HashMap;

        let mut result = i32::min_value();
        let mut sum = 0;
        let mut visits = HashMap::<i32, usize>::new();
        let mut left = 0usize;
        let mut right = 0usize;
        let nums_len = nums.len();
        while right < nums_len {
            let rn = nums[right];
            if let Some(visited) = visits.get(&rn) {
                while left <= *visited {
                    sum -= nums[left];
                    left += 1;
                }
            }

            sum += rn;
            visits.insert(rn, right);
            result = result.max(sum);
            right += 1;
        }

        result
    }
}
